### ISSUE TITLE
Bug: [no-unnecessary-condition] doesn't report on unnecessary optional array index access when `noUncheckedIndexedAccess` is enabled

### ISSUE THREAD

**Issue by ronami** [typescript-eslint/typescript-eslint#10960]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.8.2&fileType=.tsx&code=CYUwxgNghgTiAEYD2A7AzgF3hkmBc8AgjDFAJ4A8A3vFAPwGYwCWKA5vAL4B8A3AFD8A9EPgAlEAAckMHMHgAjMvAAqZSSADKYFpKxQ02ABYIkCgFbgszQ9LRpmCiMoDkAVxSgAZqxDAX-DiYANoADAC6AHRQAvzMXvAAFEEYYeF00QCU8FT88PAi8ADyesyoUBCIRlCsrBw2KC5YKOC4aLDKJnB52LipEdG8%2Bfn8nILxSSlpGVDZufmFmkZIbhDyCghw0rJ%2Bim5YwEi4jRh0PVMRMwJjQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3QAacDUhFYAVSLQ%2Bg9ABMAkkVXoAHmoCC0aOhLCwo9FIC%2BIK0A&tokens=false

### Repro Code

```TypeScript
declare const test: Array<{ a?: string }>;

// Reported by TypeScript as the object is possibly 'undefined'
test[0].a;

if (test[0]?.a) {
  // Optional chaining isn't necessary here
  test[0].a;    
}

if (test[0]?.a) {
  // Should be reported but doesn't?
  test[0]?.a;
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-condition": "error",
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true,
    "noUncheckedIndexedAccess": true
  }
}
```

### Expected Result

I expected the rule to report on unnecessary optional array index access when `noUncheckedIndexedAccess` is enabled.

### Actual Result

The rule seems to fails to catch these unnecessary optional array index access cases when `noUncheckedIndexedAccess` is enabled.

**Comment by kirkwaiblinger** [comment]
I'm guessing it's because the array index exception applies unconditionally, and doesn't take into account that the optional chain could be "actually unnecessary". Maybe we could disable that line when `noUncheckedIndexedAccess` is enabled?

https://github.com/typescript-eslint/typescript-eslint/blob/dba0bed86fd031337d28405d5ee0c2e6ca4718de/packages/eslint-plugin/src/rules/no-unnecessary-condition.ts#L357-L362

**Comment by ronami** [comment]
> I'm guessing it's because the array index exception applies unconditionally, and doesn't take into account that the optional chain could be "actually unnecessary" for CFA reasons. Maybe we could disable that line when `noUncheckedIndexedAccess` is enabled?
> 
> [typescript-eslint/packages/eslint-plugin/src/rules/no-unnecessary-condition.ts](https://github.com/typescript-eslint/typescript-eslint/blob/dba0bed86fd031337d28405d5ee0c2e6ca4718de/packages/eslint-plugin/src/rules/no-unnecessary-condition.ts#L357-L362)
> 
> Lines 357 to 362 in [dba0bed](/typescript-eslint/typescript-eslint/commit/dba0bed86fd031337d28405d5ee0c2e6ca4718de)
> 
>  // Since typescript array index signature types don't represent the 
>  //  possibility of out-of-bounds access, if we're indexing into an array 
>  //  just skip the check, to avoid false positives 
>  if (isArrayIndexExpression(expression)) { 
>    return; 
>  }

Yes! That was my approach trying to tackle this locally, seems rather trivial and doesn't fail any existing test 👍

**Comment by kirkwaiblinger** [comment]
🚀
